### DISCUSSION TITLE
Google Rich Snippets extensions

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#4299] [Questions & Answers]
I am looking for an extension to implement **Google Rich Snippets** based on this criteria

- it should be full featured with a good UI
- it should be maintained (if possible)
- reasonable price (if paid)

If your are using one already please share your opinions. It doesn't matter if the extension if free or not.

Here is a list for paid extensions after searching the Internet

https://www.mageworx.com/seo-suite-ultimate-magento-extension.html
https://mirasvit.com/magento-extensions/advanced-seo-suite.html
https://aheadworks.com/magento-1-extensions/ultimate-seo-suite-extension-for-magento-1
https://www.wyomind.com/google-rich-snippets-magento.html
https://amasty.com/magento-google-rich-snippets.html
https://amasty.com/magento-seo-toolkit.html

**Comment by sreichel** [comment]
Id go with amasty.

I know they dont maintain M1 anymore but this can be done by :dollar:  :P

**Reply by addison74** [reply]
After reading the user manuals this is the top 3 list for paid extensions.

1. MageWorx SEO Suite Extension
2. Amasty SEO Toolkit
3. Mirasvit Advanced SEO Suite

None of them are maintained. You can pay 150 bucks and you could get a nice incompatibility with newer PHP versions that need fixes.

I am looking for free extensions to evaluate ... if they are any.

**Comment by addison74** [comment]
I discussed with both companies. Amasty is offering only 15% discount for an unmaintained product. MageWorx is going up to 35%. In my opinion both extensions need intensive tests and changes to run in the latest PHP version and follow the new SEO rules. A better discount was a fair bet, around 50%. Anyway I do not want to interfere with anyone business.

I decided to give a try to MageWorx because of the discount and the good reviews over the competitions. Here is a good comparison

https://www.mageworx.com/blog/comparison-review-magento-seo-extensions

PS - I am surprised that products are offered that are no longer updated, they are no longer maintained at the original prices. These developers do not take into account that you will have to hire someone to update them for you and make them run. Hats off to those who chose to offer their paid extensions for Magento 1 for free.

**Reply by sreichel** [reply]
As said ... i'd prefered A ....

> These developers do not take into account that you will have to hire someone to update them for you and make them run.

Or ... if someone still uses M1, he knows what he does - or hires someone ;)

You saved ~$50 ... send it to me and i make it run :P

**Comment by DegrizNet** [comment]
Maybe this comes in handy.. if you want to just edit your template.
page/html/breadcrumbs.phtml

```
<div class="breadcrumbs">
    <ul itemscope itemtype="http://schema.org/BreadcrumbList">
        <?php $s = 1; foreach($crumbs as $_crumbName=>$_crumbInfo): ?>
            <li class="<?php echo $_crumbName ?>" itemprop="itemListElement" itemscope itemtype="http://schema.org/ListItem">
            <?php if($_crumbInfo['link']): ?>
                <a title="<?php echo $this->htmlEscape($_crumbInfo['title']) ?>" href="<?php echo $_crumbInfo['link'] ?>" itemprop="item"><span itemprop="name"><?php echo $this->htmlEscape($_crumbInfo['label']) ?></span></a>
				<meta itemprop="position" content="<?php echo $s; $s++; ?>" /> 
           <?php elseif($_crumbInfo['last']): ?>
                <strong itemprop="name"><?php echo $this->htmlEscape($_crumbInfo['label']) ?></strong>
				<meta itemprop="position" content="<?php echo $s; $s++; ?>" />
            <?php else: ?>
                <span itemprop="name"><<?php echo $this->htmlEscape($_crumbInfo['label']) ?></span>
				<meta itemprop="position" content="<?php echo $s; $s++; ?>" />
            <?php endif; ?>
            <?php if(!$_crumbInfo['last']): ?>
                <span>/ </span>
            <?php endif; ?>
            </li>
        <?php endforeach; ?>
    </ul>
</div>
```

catalog/product/view.phtml
```
<div class="no-display" itemprop="offerDetails" itemscope itemtype="https://schema.org/Product">
	<span itemprop="name"><?php $name = $_helper->productAttribute($_product, $_product->getName(), 'name'); echo $name; ?></span>
	<span itemprop="brand" class="no-display"><?php echo $_product->getManufacturer() ?></span>
	<span itemprop="description"><?php echo strip_tags($_helper->productAttribute($_product, $_product->getShortDescription(), 'short_description')); ?></span>
	<span itemprop="image"><?php echo  $_product->getImageUrl(); ?></span>
	<div itemprop="offers" itemscope itemtype="http://schema.org/Offer">
		<span itemprop="price"><?php echo Mage::helper('core')->currency($_product->getFinalPrice(), false, false); ?></span>
		<span itemprop="priceCurrency"><?php echo Mage::app()->getStore()->getCurrentCurrencyCode(); ?></span>
		<?php if ($_product->isAvailable()): ?>
		   <link itemprop="availability" href="http://schema.org/InStock" /><?php echo $this->__('Availability:') ?> <span><?php echo $this->__('In stock') ?></span>
		<?php else: ?>
		   <link itemprop="availability" href="http://schema.org/OutOfStock" /><?php echo $this->__('Availability:') ?> <span><?php echo $this->__('Out of stock') ?></span>
		<?php endif; ?>
        <?php 
            // Add priceValidUntil as a div
            $validUntil = date('Y-m-d', strtotime('+30 days'));
        ?>
        <div itemprop="priceValidUntil" content="<?php echo $validUntil; ?>"></div>
	</div>
</div>
```

**Reply by addison74** [reply]
Nice sharing when using Microdata format. Those extensions can do much more. OpenGraph meta's, application/ld+json format in product, category, cms pages.

More information about Rich Snippet
https://developers.google.com/search/docs/appearance/structured-data/review-snippet

**Reply by DegrizNet** [reply]
yes , i know.. for reviews you add..

```
    <?php
    // Fetching ratings and reviews
    $ratingSummary = $_product->getRatingSummary();
    $reviewsCount = Mage::getModel('review/review')
        ->getTotalReviews($_product->getId(), true);

    if ($ratingSummary && $reviewsCount) {
        $averageRating = $ratingSummary->getRatingSummary() / 20; // Magento gives rating summary out of 100, convert to 5
    } else {
        $averageRating = 5; // Default perfect score
        $reviewsCount = 1; // Default one review
    }
    ?>
    <div itemprop="aggregateRating" itemscope itemtype="http://schema.org/AggregateRating">
        <span itemprop="ratingValue"><?php echo $averageRating; ?></span>
        <span itemprop="reviewCount"><?php echo $reviewsCount; ?></span>
    </div>
```

Keep in mind that a lot of functions and data can effect performance.. so i prefer simple code over complicated module. ;)

I laso did some for category FAQ: https://github.com/DegrizNet/m1_faqschema

**Comment by addison74** [comment]
For the moment I have given up using one of the solutions above. After reading almost everything on this subject, I created my own version editing it in the phtml files.

- Open Graph, Twitter, Facebook tags went to the template/page/html/head.phtml file
- The Schema.org tags went to the /template/catalog/product/view.phtml file

I have considered several scenarios, including when metas like name, description and keywords are filled up in the product edit page, which have priority over the name, short description which are used by default. In addition, I solved a problem that I have not addressed for a long time. The ideal og:image must have 1200 x 630 px for "perfection". Well, thanks to the functions in the OpenMage, I created an image with this size and now Facebook, Twitter, Whatsapp, no longer truncate my square product images, so that the top and bottom parts were not visible. I used 

```php
$productImage = Mage::helper('catalog/image')->init($product, 'image')->resize(1200, 630);
```

The square shape product image becomes 630px height and it still has 630px width, but with white on both sides. This debugger https://imgsrc.io/tools/open-graph-debugger and the one from Facebook showed me the flawless image.

My opinion is that at least for the products we should implement out of the box Open Graph, Twitter and Schema.org. This is a must have for SEO and posting/sharing in the social medial. Some value such as the processing time of an order, the return time period and other tags can be set in the Backend.

**Reply by DegrizNet** [reply]
Yes, I did the same. My old example if somebody needs it:

```
<?php if(Mage::registry('current_product')): ?>
<?php $product = Mage::registry('current_product'); ?>

<meta property="og:title" content="<?php $ime = $product->getName(); echo $ime; ?>" />
<meta property="og:type" content="product" />
<meta property="og:image" content="<?php $slika = $this->helper('catalog/image')->init($product, 'small_image')->resize(1200,630); echo $slika; ?>" />
<meta property="og:url" content="<?php echo Mage::registry('product')->getProductUrl(); ?>" />
<meta property="og:description" content="<?php $opis = strip_tags(($product->getShortDescription())); echo  $opis; ?>" />
<meta property="og:site_name" content="<?php echo Mage::app()->getWebsite()->getName(); ?>" />
<meta property="og:price:amount" content="<?php $cena = Mage::registry('product')->getFinalPrice(); echo $cena; ?>" />
<meta property="og:price:currency" content="<?php $currency = Mage::app()->getStore()->getCurrentCurrencyCode(); echo $currency; ?>" />

<?php elseif(Mage::registry('current_category')): ?>

<meta property="og:title" content="<?php echo $this->getTitle() ?>" />
<meta property="og:type" content="product.group" />
<meta property="og:url" content="<?php echo $this->helper('core/url')->getCurrentUrl();?>" />
<meta property="og:description" content="<?php echo strip_tags($this->getDescription()) ?>" />
<meta property="og:site_name" content="<?php echo Mage::app()->getWebsite()->getName(); ?>" />
<?php elseif ($post = Mage::registry('wordpress_post')): ?>

<meta property="og:title" content="<?php echo $post->getData('post_title'); ?>" />
<meta property="og:type" content="article" />
<meta property="og:image" content="<?php if ($featuredImage = $post->getFeaturedImage()) { echo $featuredImage->getAvailableImage(); } ?>" />
<meta property="og:url" content="<?php echo $this->helper('core/url')->getCurrentUrl();?>" />
<meta property="og:description" content="<?php echo substr(strip_tags($post->getData('post_content')), 0, 140).'..';   ?>" />
<meta property="og:site_name" content="<?php echo Mage::app()->getWebsite()->getName(); ?>" />

<?php elseif((Mage::getSingleton('cms/page')->getIdentifier() == 'home' && Mage::app()->getFrontController()->getRequest()->getRouteName() == 'cms')) : ?>

<meta property="og:title" content="<?php echo $this->getTitle() ?>" />
<meta property="og:type" content="website" />
<meta property="og:url" content="<?php echo $this->helper('core/url')->getCurrentUrl();?>" />
<meta property="og:description" content="<?php echo strip_tags($this->getDescription()) ?>" />
<meta property="og:site_name" content="<?php echo Mage::app()->getWebsite()->getName(); ?>" />

<?php else: ?>
<meta property="og:title" content="<?php echo $this->getTitle() ?>" />
<meta property="og:type" content="article" />
<meta property="og:url" content="<?php echo $this->helper('core/url')->getCurrentUrl();?>" />
<meta property="og:description" content="<?php echo strip_tags($this->getDescription()) ?>" />
<meta property="og:site_name" content="<?php echo Mage::app()->getWebsite()->getName(); ?>" />

<?php endif; `?>
```

**Reply by addison74** [reply]
It is almost the same like my implementation. I would take in consideration meta title, meta description having priority over name and short description if these values are set in the Backend. In my case short_description is different than meta_description, the last one is limited to max 160 characters, the same for meta_title which is limited to max 60 characters. I don't remember if we implemented or not in OpenMage the counting of characters for the SEO fields.

**Reply by DegrizNet** [reply]
> It is almost the same like my implementation. I would take in consideration meta title, meta description having priority over name and short description if these values are set in the Backend. In my case short_description is different than meta_description, the last one is limited to max 160 characters, the same for meta_title which is limited to max 60 characters. I don't remember if we implemented or not in OpenMage the counting of characters for the SEO fields.

yes this should probably need to update.. like change substr to mb_substr (to not cut some special characters) etc. ..I pasted old code from my M1 docs.
